File: app/src/main/java/com/zionhuang/music/youtube/RetrofitManager.java
package com.zionhuang.music.youtube;

import com.google.gson.GsonBuilder;

import retrofit2.Retrofit;
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitManager {
    private static final String GoogleSuggestionAPIBaseUrl = "https://clients1.google.com/";
    private static volatile RetrofitManager mInstance;
    private SuggestionAPIService mSuggestionAPIService;

    private RetrofitManager() {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(GoogleSuggestionAPIBaseUrl)
                .addConverterFactory(GsonConverterFactory.create(new GsonBuilder()
                        .registerTypeAdapter(SuggestionResult.class, SuggestionResult.deserializer)
                        .create()))
                .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
                .build();
        mSuggestionAPIService = retrofit.create(SuggestionAPIService.class);
    }

    public static RetrofitManager getInstance() {
        if (mInstance == null) {
            synchronized (RetrofitManager.class) {
                if (mInstance == null) {
                    mInstance = new RetrofitManager();
                }
            }
        }
        return mInstance;
    }

    public SuggestionAPIService getSuggestionAPIService() {
        return mSuggestionAPIService;
    }
}

File: app/src/main/java/com/zionhuang/music/utils/RetrofitManager.kt
package com.zionhuang.music.utils

import com.google.gson.GsonBuilder
import com.jakewharton.retrofit2.adapter.kotlin.coroutines.CoroutineCallAdapterFactory
import com.zionhuang.music.youtube.api.YouTubeSuggestionAPIService
import com.zionhuang.music.youtube.models.YouTubeSuggestion
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object RetrofitManager {
    private const val GoogleSuggestionAPIBaseUrl = "https://clients1.google.com/"

    private val retrofit: Retrofit = Retrofit.Builder()
            .baseUrl(GoogleSuggestionAPIBaseUrl)
            .addConverterFactory(GsonConverterFactory.create(GsonBuilder()
                    .registerTypeAdapter(YouTubeSuggestion::class.java, YouTubeSuggestion.deserializer)
                    .create()))
            .addCallAdapterFactory(CoroutineCallAdapterFactory())
            .build()

    val suggestionAPIService: YouTubeSuggestionAPIService = retrofit.create(YouTubeSuggestionAPIService::class.java)
}